lib/pavilion/utils.py

```diff
@@ -405,6 +405,80 @@ def union_dictionary(dict1, dict2):
 
     return dict1
 
+def flatten_nested_dict(dict_in, keycollect='', new_d=dict(), keysplit='.'):
+    """ Takes a nested dictionary and concatenates its nested keys
+    in a single key at the top level.
+
+    Remove keys whose value evaluates to False.
+    If dict_in has only one key and it's not the final key, discard it.
+    Remove key concatenator from front/back of key string.
+    """
+
+    knew = keycollect
+    for key, val in dict_in.items():
+        if not val:
+            continue
+        if len(dict_in.keys()) > 1 or not isinstance(val, dict):
+            knew = keysplit.join([keycollect,key])
+        if not isinstance(val, dict):
+            knew = knew.strip(keysplit)
+            new_d[knew] = val
+        else:
+            new_d = flatten_nested_dict(val, knew, new_d.copy())
+
+    return new_d
+
+
+def flatten_dictionaries(nested_dicts):
+    """ Takes a (possibly nested) dictionary or list of dictionaries
+    and returns a dictionary that contains only (key: value) pairs
+    by merging nested keys.
+
+    Remove keys whose value evaluates to False.
+    Returns the type you gave it, list or dict.
+    When when flattening a sub dictionary, check that the resulting
+    keys are not already in the main dict to be overwritten when
+    the flattened key:values are added. If the key is present,
+    append the parent dictionary key referring to the sub dictionary
+    to the flattened key and add to the flattened dictionary.
+    """
```

Comment:
I don't think you're right about any of these from a techno-philosophical standpoint. 

1. If I'm not mistaken, the value can be a list but that list can't have dictionaries.  So your example would be impossible.  It could be `somekey.subkey.subsubkey.0` or even nested lists 0.1 and those lists could be different sizes.  That's the way it looks from the way the yaml keys are structured (if you hand sched a list in any key it will throw an error).  But my idea here is just to make non dictionary values presentable. So currently `somekey.0.subkey.*.subsubkey` is in accessible as it was before but `somekey.subkey` is accessible when it wasn't before. Of course, you recognized that it's an improvement not a complete solution, I'm just not sure that the problem you described is possible.  

2.  REDACTED

3. I can see how you could just take the keys and smash them together and then break them apart again to recover the values.  That makes sense, but you'd need to traverse the tree twice, once to get the keys and concatenate them and then to get the values.  In this way you just pick up the value and the key when you get there. I think this way would scale better.

Reply:
I changed it to just take dicts and named it flatten_dictionary.  Now the list of flattened dicts is collected in a for loop results.py.